src/05-run.md

```diff
@@ -0,0 +1,164 @@
+we exit the run loop. Next we update our `socketwrappermap` by looping over all of the values in `threadswaitingfor`
+and insert each `recvr` and `sendr` into the key of `skt.inner_sock`.
+
+With all the book keeping done, we run all of the `timers` that have reached their deadline, and update our
+local variable `timeout` to the duration until the shortest remaining timeout. If at least one thread was added to
+`readythreads`, we set our timeout to 0, because we already know we have new work to do. At this point if `timeout` is
+`nil` and both `sendt` and `recvt` are empty, we raise an error because we are about to call
+`socket.select({}, {}, nil)` which would just block forever. At this point, we call `socket.select` capturing the
+values in `recvr`, `sendr` and `err`. If `err` isn't the value `"timeout"`, we raise that error. If `err` is `nil`
+or `"timeout"`, we loop over all of the values in `recvr` and `sendr`, looking up the cosock socket in
+`socketwrappermap` and calling `skt:_wake` which would call the function we provided to `setwaker` above.
+
+With that complete, we now have a fully updated `readythreads` and we start the loop again.
+
+<details>
+<summary>Outline Version</summary>
+
+1. Define `wakethreads`
+2. Define an empty list of senders (`sendt`), receivers (`recvt`) and a `timeout`
+3. Pop all `readythreads` entries into the `wakethreads`
+4. Loop over all threads in `wakethreads`
+   1. If `coroutine.status` for that thread returns "suspended"
+      1. Clear any timers
+      2. Clear any wakers registered with a `timeout`
+      3. `coroutine.resume` with the stored `recv4`, `sendr` and `err` arguments
+      4. If `coroutine.resume` returned `true` in the first position and `coroutine.status` returns "suspended"
+         1. Re-populate `threadswaitingfor[thread]` with the 3 other return values from `coroutine.resume`
+            1. These should be the `recvt`, `sendt` and `timeout` values that will populate select args
+         2. Set the waker for all sockets in `recvt` and `sendt` to call `wake_thread` and then unset themselves
+         3. If `coroutine.resume` returned a `timeout`, create a new timer for this thread which will call `wake_thread_err` on expirations with the value "timeout"
+      5. if `coroutine.status` returned "dead"
+         1. If `coroutine.resume` returned `false` in the first position and no `threaderrorhandler` has been set
+            1. Raise an error
+               1. If the `debug` library is available, include a `debug.traceback` and the second return value from `cosock.resume`
+               2. Else just raise an error with the second return value from `cosock.resume`
+            2. Exit the application
+               1. This calls `os.exit(-1)`
+   2. Else, print a warning message if printing is turned on
+5. Initialize a variable `running` to `false`
+6. Loop over all `threads`, if `coroutine.status` doesn't return "dead" for any, set `running` to `true`
```

Comment:
Is this s/any/all?

Reply:
Here I've re-worked it and I think it is a bit clearer.

```
Loop over all `threads`, calling `coroutine.status` on each, if at least 1 doesn't return "dead", set `running` to `true`
```